### PR TITLE
Core: Partition filter pushdown for entries table

### PR DESCRIPTION
**Pull request by szehon-ho**
This implements partition filter pushdown for 'entries' and 'all_entries' table. This makes a base class for both and then re-uses the partition filter pushdown logic of 'files' table when doing planning.

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/BaseEntriesTable.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  }
+
+  @Override
+  public Schema schema() {
+    StructType partitionType = Partitioning.partitionType(table());
+    Schema schema = ManifestEntry.getSchema(partitionType);
+    if (partitionType.fields().size() < 1) {
+      // avoid returning an empty struct, which is not always supported. instead, drop the partition
+      // field (id 102)
+      return TypeUtil.selectNot(schema, Sets.newHashSet(102));
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseEntriesTable.java:53]: Can you use `DataFile.PARTITION_ID` here? We hard-coded 102 before that was a constant.

File: core/src/main/java/org/apache/iceberg/BaseEntriesTable.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    return CloseableIterable.transform(
+        filteredManifests,
+        manifest ->
+            new ManifestReadTask(
+                table, manifest, projectedSchema, schemaString, specString, residuals));
+  }
+
+  @Override
+  MetadataTableType metadataTableType() {
+    return MetadataTableType.ALL_ENTRIES;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/BaseEntriesTable.java:96]: This should be abstract, right?

File: core/src/test/java/org/apache/iceberg/TestMetadataTableFilters.java [hunk 3]
```diff
@@ -153,20 +160,46 @@ private boolean isAggFileTable(MetadataTableType tableType) {
...
-                    102,
-                    "partition",
-                    Types.StructType.of(optional(1000, "data_bucket", Types.IntegerType.get())),
-                    "Partition data tuple, schema based on the partition spec"))
-            .asStruct();
-
-    TableScan scan = metadataTable.newScan().select("partition.data_bucket");
-    Assert.assertEquals(expected, scan.schema().asStruct());
+
+    TableScan scan = metadataTable.newScan().select(partitionColumn("data_bucket"));
```
Comment (rdblue) [inline core/src/test/java/org/apache/iceberg/TestMetadataTableFilters.java:202]: Why remove the assertion? Is it not useful?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Looks good overall, just a couple of minor things you may want to look at.

File: core/src/main/java/org/apache/iceberg/BaseEntriesTable.java (hunk 2 above)
Comment (szehon-ho) [reply core/src/main/java/org/apache/iceberg/BaseEntriesTable.java]: Whoops you are right, thanks

File: core/src/test/java/org/apache/iceberg/TestMetadataTableFilters.java (hunk 3 above)
Comment (szehon-ho) [reply core/src/test/java/org/apache/iceberg/TestMetadataTableFilters.java:202]: Ah I wrote this Test class initially to have one partition-filter-pushdown test various metadata tables (files, all_files, etc), but this assertion would kind of not work here.  There should be other tests that assert the same thing (can add if not), and also end to end sql tests that excercise this.

**Comment by szehon-ho** [comment]
Thanks  @rdblue , for the fast review

**Comment by aokolnychyi** [comment]
LGTM too

**Pull request merged**